_overviews/scala3-book/collections-methods.md

```diff
@@ -0,0 +1,310 @@
+```
+
+The REPL shows the result:
+
+```scala
+scala> oneToTen.filter(_ < 4).foreach(println)
+1
+2
+3
+```
+
+
+
+## `head`
+
+The `head` method comes from Lisp and other earlier functional programming languages. It’s used to print the first element (the head element) of a list:
+
+```scala
+oneToTen.head   // Int = 1
+names.head      // adam
+```
+
+Because a `String` is a sequence of characters, you can also treat it like a list. This is how `head` works on these strings:
+
+```scala
+"foo".head   // f
+"bar".head   // b
+```
+
+`head` is a great method to work with, but as a word of caution it can also throw an exception when called on an empty collection:
+
+```scala
+val emptyList = List[Int]()   // emptyList: List[Int] = List()
+emptyList.head                // java.util.NoSuchElementException: head of empty list
+```
+
+Because of this you may want to use `headOption` instead of `head`, especially when programming in a functional style:
+
+```scala
+emptyList.headOption          // Option[Int] = None
+```
+
+As shown, it doesn’t throw an exception, it simply returns the type `Option` that has the value `None`. You can learn more about this programming style in the Reference documentation.
+
+
+
+## `tail`
+
+The `tail` method also comes from Lisp, and it’s used to print every element in a list after the head element. A few examples demonstrate this:
+
+```scala
+oneToTen.head   // Int = 1
+oneToTen.tail   // List(2, 3, 4, 5, 6, 7, 8, 9, 10)
+
+names.head      // adam
+names.tail      // List(brandy, chris, david)
+```
+
+Just like `head`, `tail` also works on strings:
+
+```scala
+"foo".tail   // "oo"
+"bar".tail   // "ar"
+```
+
+Like `head` and `headOption`, there’s also a `tailOption` method, which is preferred in functional programming.
+
+
+
+## `take`, `takeRight`, `takeWhile`
+
+The `take`, `takeRight`, and `takeWhile` methods give you a nice way of “taking” the elements from a list that you want to use to create a new list. This is `take` and `takeRight`:
+
+```scala
+oneToTen.take(1)        // List(1)
+oneToTen.take(2)        // List(1, 2)
+
+oneToTen.takeRight(1)   // List(10)
+oneToTen.takeRight(2)   // List(9, 10)
+```
+
+And this is `takeWhile`, which works with a predicate function:
+
+```scala
+oneToTen.takeWhile(_ < 5)       // List(1, 2, 3, 4)
+names.takeWhile(_.length < 5)   // List(adam)
+```
+
+
+
+## `drop`, `dropRight`, `dropWhile`
+
+`drop`, `dropRight`, and `dropWhile` are essentially the opposite of their “take” counterparts, dropping elements from a list. Here are some examples:
+
+```scala
+oneToTen.drop(1)        // List(2, 3, 4, 5, 6, 7, 8, 9, 10)
+oneToTen.drop(5)        // List(6, 7, 8, 9, 10)
+
+oneToTen.dropRight(8)   // List(1, 2)
+oneToTen.dropRight(7)   // List(1, 2, 3)
+```
+
+And this is `dropWhile`, which works with a predicate function:
+
+```scala
+oneToTen.dropWhile(_ < 5)       // List(5, 6, 7, 8, 9, 10)
+names.dropWhile(_ != "chris")   // List(chris, david)
+```
+
```

Comment:
I think it is useful to show the behavior of `drop*` on corner cases, in this case when you ask `n` elements where `n > seq.size`